File: 242-ValidAnagram/242-ValidAnagram.java
// Last updated: 11/2/2025, 8:23:24 PM
class Solution {
    public boolean isAnagram(String s, String t) {
        if(s.length() != t.length()){
            return false;
        }
        //Using a Counter
        /*
        Creates a size 26 int arrays as buckets for each letter in alphabet. 
        It increments the bucket value with String s and decrement with string t. 
        So if they are anagrams, all buckets should remain with initial value which is zero. 
        Checks if all the buckets are 0 at the end, if it is true, if not it's false
        */

        int[] store = new int[26];

        for (int i = 0; i < s.length(); i++) {
            //'a' just gives you distance between 'a' and the character which is nothing but index value of freq array
            //'a' = 0, 'b' = 1, 'c' = 2.... 
            store[s.charAt(i) - 'a']++; 
            store[t.charAt(i) - 'a']--; 
        }

        for (int n : store){
            if (n != 0){
                return false;
            }
        }

        return true;
    }
}

//Time Complexity: O(n), accessing the counter table is a constant time operation 
//Space complexity: O(1). space complexity is O(1) because the table's size stays constant no matter how large n is.

File: 242-ValidAnagram/242-ValidAnagram.py
# Last updated: 11/2/2025, 8:23:22 PM
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False #if the strings aren't the same size, can't be an anagram

        #hashmap for each string
        hashmapS = {} 
        hashmapT = {} 
        
        #can use s or t, since we know now it is the same length
        for i in range(len(s)):
            hashmapS[s[i]] = 1 + hashmapS.get(s[i],0)
            hashmapT[t[i]] = 1 + hashmapT.get(t[i],0)
        
        #check if the hashmaps are the same
        #if they are that means it's an anagram and returns true, false otherwise
        return hashmapS == hashmapT

        #Time Complexity:O(n + m),n is the length of string s and m is the length of string t
        #Space Complexity:O(1), since we have at most 26 different characters.
            